TypeScript starter kit

stoianchoo opened the pull request:
I have created a TypeScript translation of the JavaScript starter kit.

lidavidm commented on starter_kits/TypeScript/README.md:
```diff
@@ -0,0 +1,6 @@
+Run `npm install` to get the typescript definitions for node.
+
+# Development
+Don't forget to run `tsc` to compile the typescript sources before using the bot.
```
Is it possible to add this step to `run_game.sh`/`run_game.bat`?

stoianchoo replied:
Yes. I have updated my fork with the changes. What't the next step?

lidavidm commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,317 @@
+    structure: any = null;
+
+    constructor(
+        public position: Position,
```
Didn't know about this feature of TS, that's pretty cool.

stoianchoo replied:
Yup, many cool things in TS. Eliminates a lot of the errors of working with JS. Good for new developers, great support in the IDEs

stoianchoo commented:
I was thinking if it is better to extract all of the Classes in their own files?

lidavidm commented:
@stoianchoo up to you! It works either way.

fqlx commented on starter_kits/TypeScript/tsconfig.json:
```diff
@@ -0,0 +1,60 @@
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
+
+    /* Additional Checks */
+    // "noUnusedLocals": true,                /* Report errors on unused locals. */
```
I'd enable all these strictness checks.  Or at least enable it for this code to clean up and then comment them out for beginners.

stoianchoo replied:
Enabled all checks. I thinks it's good even for beginners as it prevents subtle errors.

stoianchoo commented:
@fqlx Thanks for the feedback. I tried to keep it as close as possible to the JS kit, but you are right, I'll try to bring it up to spec.

stoianchoo commented on starter_kits/TypeScript/MyBot.ts:
```diff
@@ -0,0 +1,49 @@
+            }
+        }
+
+        if (game.turnNumber < 0.75 * Constants.MAX_TURNS &&
```
Removed all magic numbers in the main file

stoianchoo commented on starter_kits/TypeScript/MyBot.ts:
```diff
@@ -0,0 +1,49 @@
+    // At this point "game" variable is populated with initial map data.
+    // This is a good place to do computationally expensive start-up pre-processing.
+    // As soon as you call "ready" function below, the 2 second per turn timer will start.
+    await game.ready('MyJavaScriptBot');
```
Done

stoianchoo commented on starter_kits/TypeScript/run_game.bat:
```diff
@@ -0,0 +1 @@
+tsc && halite.exe --replay-directory replays/ -vvv --width 32 --height 32 "node .\out\MyBot.js" "node .\out\MyBot.js"
```
I think I can't. Maybe https://github.com/TypeStrong/ts-node but it's another dependency and maybe another can of worms.

fqlx replied:
Ah okay.

stoianchoo commented on starter_kits/TypeScript/hlt/entity.ts:
```diff
@@ -0,0 +1,92 @@
+     * Return a command to move this ship in a direction without
+     * checking for collisions.
+     */
+    move(direction: string | Direction) {
```
Yup, makes it more resilient.

stoianchoo replied:
Yep. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,317 @@
+
+/** A cell on the game map. */
+export class MapCell {
+    ship: Ship | null = null;
```
The link you mentioned states:"AGAIN: This is NOT a prescriptive guideline for the TypeScript community", but I get your point. It's basically checked internally so it can go either way.
Changed.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,317 @@
+    /**
+     * Update the player object for the current turn using input from
+     * the game engine.
+     * @private
```
Hmm. Not sure how this got here. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,317 @@
+    /** Get a list of the player's dropoffs. */
+    getDropoffs() {
+        const result = [];
+        for (const dropoff of this._dropoffs.values()) {
```
They basically want to represent dropoffs as an array as Map lacks a lot of Arrays functional methods. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,317 @@
+    getShips() {
+        const result = [];
+        for (const ship of this._ships.values()) {
+            result.push(ship);
```
They basically want to represent ships as an array as Map lacks a lot of Arrays functional methods. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/entity.ts:
```diff
@@ -0,0 +1,92 @@
+
+/** Base entity class for Ships, Dropoffs, and Shipyards. */
+export class Entity {
+    constructor(public owner: any, public id: number, public position: Position) {
```
Yes. I checked, it is used as a number. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/logging.ts:
```diff
@@ -0,0 +1,30 @@
+    }
+    
+    static format(args: any) {
+        return args.map((x:any) => typeof x !== 'string' ? x.toString() : x).join(' ');
```
Not sure what you mean as not supported. "Every object has a toString()" https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toString

fqlx replied:
Ah good point. No issue here.

stoianchoo commented:
@lidavidm I started implementing @fqlx suggestions at 8fae5ba34cd6218adfd6ca292a44162e632f753e. If you don't approve, maybe they can be reverted, although I think they are good improvements.
I think we can improve the code more. 
* For example all stdin line parsing should be moved to a separate class.

lidavidm commented:
@stoianchoo great, happy to hold off until you're satisfied with the kit. I think adding more checks to `ship.move` might make it too inconsistent with other kits, though.

stoianchoo commented:
Exactly. That is the question. Is keeping the code similar to other kits more important than code quality of the specific kit? I don't know what the reason is for keeping them the same. But I don't think the kit should generate errors (in case of move without halite check, also it's a warning but I think it should be an error as it represents a discrepancy from where the bot thinks it is and the actual position).
Anyway I assume the game target is the high level algorithm development so this low level stuff is just in the way and can demotivate beginners.

lidavidm commented:
Broadly, we'd like to keep them similar. Adding things like unit tests or making ship instances persistent is perfectly fine (especially as we'd like to add the latter feature to all kits). 

Generating an error on movement, however, would be drastically different than other kits, especially as most logic you'll see does not check for this.

stoianchoo commented:
@lidavidm Generating an error in response of a movement command without enough halite was for the game engine not for the kits. I think it will promote better interface between the bot and the game server. It will avoid a common class of errors.
Maybe I should create a separate issue like "Improvements to the kit's and the game engine"?

stoianchoo commented:
@fqlx Do you have input on the two unresolved conversations?

lidavidm commented:
@stoianchoo ah, I see. I feel it would be too punishing for users to kick them out if they try to move a ship without enough energy; we had discussions about this before launch. Also, at this point, it would break most peoples' bots, so it's not a change we would want to make. 

The STRICT_ERRORS flag in the engine might be what you want if you want this personally, though.

fqlx commented on starter_kits/TypeScript/hlt/Game.ts:
```diff
@@ -0,0 +1,140 @@
+    _getLine: () => Promise<string>;
+    public myId: number = 0;
+    public players = new Map<number, Player>();
+    public me: Player | undefined = undefined;
```
You can do
`me ?: Player`

You don’t have to explicitly have `undefined` with the `?` operator.

stoianchoo replied:
Good point. I'll update all.

fqlx commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,183 @@
+     * @returns A normalized position object fitting within the bounds of the map
+    */
+    normalize(position: Position) {
+        let x = ((position.x % this.width) + this.width) % this.width;
```
Use `const`

stoianchoo replied:
Done

fqlx commented on starter_kits/TypeScript/hlt/Direction.ts:
```diff
@@ -0,0 +1,64 @@
+        else if (this.equals(Direction.Still)) {
+            return Direction.Still;
+        }
+        throw new Error(`Non-cardinal direction cannot be inverted: ${this}`);
```
Log the Direction here

fqlx replied:
No else?

stoianchoo replied:
* They log the direction with `${this}` in the error string
* else is omitted because of the returns in the if statements
I would say that Direction class should represent only the cardinal directions. This is supported by it's usage in the code as signifying a possible next move.

fqlx replied:
Ah okay.  I personally prefer to explicitly use `else` but I know others don't like it (because it is explicit).  We are good here.

fqlx commented on starter_kits/TypeScript/MyBot.ts:
```diff
@@ -0,0 +1,64 @@
+    while (true) {
+        await game.updateFrame();
+
+        const gameMap =  game.gameMap;
```
Extra space after the equals

stoianchoo replied:
Fixed.

fqlx commented on starter_kits/TypeScript/hlt/MapCell.ts:
```diff
@@ -0,0 +1,54 @@
+export class MapCell {
+    ship: Ship | undefined = undefined;
+    structure: Shipyard | Dropoff | undefined = undefined;
+    constructor(public position: Position, public haliteAmount: number) {
```
Empty constructor, delete it?

stoianchoo replied:
With the `public` modifiers before the arguments this expression defines and sets them as members.

fqlx replied:
Didn't know that. Awesome.

fqlx commented on starter_kits/TypeScript/hlt/Game.ts:
```diff
@@ -0,0 +1,140 @@
+        if(this.gameMap instanceof GameMap) {
+            await (this.gameMap)._update(this._getLine);
+        } else {
+            throw "Game.updateFrame() called when Game is not initialized!";
```
Log gameMap

stoianchoo replied:
Refactored

fqlx commented on starter_kits/TypeScript/hlt/Random.ts:
```diff
@@ -0,0 +1,20 @@
+    // Returns a PRN in range [0,1)
+    next() {
+        this._seed = this._seed * 16807 % 2147483647; // PRN in [1,2147483646].
+        return (this._seed - 1) / 2147483646;
```
Wtf? Can you wrap the RNG in a function.  Wrap non obvious code in a function with a descriptive name.

fqlx commented on starter_kits/TypeScript/hlt/Game.ts:
```diff
@@ -0,0 +1,140 @@
+                .split(/\s+/)
+                .map(x => parseInt(x, 10));
+            const player = this.players.get(playerId);
+            if(player instanceof Player) {
```
I don’t see a reason to type check if we are going to crash anyways

stoianchoo replied:
Refactored to a cast.

stoianchoo commented:
Thanks for the great feedback. I have implemented "Persistent ships" and moved all of the bot-server communication in one place in Game. Not committed it yet because I think I have a bug. Hard to say - no unit tests, but now as the bot has seeded RNG I look at how much halite it collects on a specific map. I think I'll create a separate communication class and leave Game trimmed down.
I'll have a look at your remarks tomorrow.

lidavidm commented on starter_kits/TypeScript/README.md:
```diff
@@ -0,0 +1,6 @@
+# Setup
+You should have typescript installed before developing. You can do this globally with `npm install -g typescript`
```
Yarn vs NPM is more of a personal choice.

stoianchoo replied:
Now can install and use local typescript and tslint.

stoianchoo commented on starter_kits/TypeScript/run_game.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+tsc && ./halite --replay-directory replays/ -vvv --width 32 --height 32 "node MyBot.js" "node MyBot.js"
```
This seems more like IDE and development preference. I'm trying to do minimum necessary setup.

fqlx replied:
I love linting and everyone should lint their code.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,183 @@
+    static _getTargetDirection(source: Position, target: Position): [Direction | null, Direction | null] {
+        return [
+            target.y > source.y ? Direction.South :
+                (target.y < source.y ? Direction.North : null),
```
Yup. Fixed.

stoianchoo commented on starter_kits/TypeScript/hlt/Game.ts:
```diff
@@ -0,0 +1,140 @@
+                while (buffer.length === 0) {
+                    await currentPromise;
+                }
+                resolve(<string>buffer.shift()); // With the above condition we guarantie that we return a sting
```
Rewritten.

stoianchoo commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,183 @@
+     * @param location the position or entity to access in this map
+     * @returns the contents housing that cell or entity
+    */
+    get(location: Location | Position | Entity): MapCell {
```
Removed the Location variant, but it's just how we generally do overloads.

stoianchoo commented:
Besides the other improvements I have extracted the async server communication to it's own class. Thus:
* Separates concerns and simplifies the code.
* Makes it for the rest of the bot code as if the server talked JSON.

If the server communicated with JSON it would be easier for users to:
* improve testing via mocking the server
* easy to add internal server to allow for planning and/or gather data for AI training

fqlx commented on starter_kits/TypeScript/hlt/ServerCommunicaion.ts:
```diff
@@ -0,0 +1,117 @@
+            data.players.push({ id, x, y });
+        });
+
+        const [, mapHeight] = await this.getNumbers();
```
I'd add the underscore as blackhole var here (if it doesn't throw an error).
`const [_, mapHeight]`

stoianchoo replied:
It does throw an error for me: `hlt/ServerCommunicaion.ts:85:16 - error TS6133: '_' is declared but its value is never read.`

fqlx replied:
Ah okay.  I'm not surprised.

fqlx commented on starter_kits/TypeScript/hlt/gameMap.ts:
```diff
@@ -0,0 +1,141 @@
+ */
+export class GameMap {
+    readonly cells: MapCell[][];
+    public width: number = 0;
```
Not a big issue, but defining `= 0` will implicitly set `number` type. So we could delete `:number` everywhere.

stoianchoo replied:
Removed everywhere.

lidavidm commented:
Thanks @stoianchoo for all the work here and thanks @fqlx for reviewing!

stoianchoo commented:
@lidavidm  Anytime :) Feel free to reach out if anything else needs fixing or improving on this.
